python/ray/util/spark/utils.py

```diff
@@ -0,0 +1,411 @@
+    """
+    Return the available heap memory and object store memory for each ray worker.
+    NB: We have one ray node per spark task.
+    """
+    num_cpus_per_spark_task = int(
+        spark.sparkContext.getConf().get("spark.task.cpus", "1")
+    )
+
+    def mapper(_):
+        try:
+            num_cpus = _get_cpu_cores()
+            num_task_slots = num_cpus // num_cpus_per_spark_task
+
+            physical_mem_bytes = _get_total_physical_memory()
+            shared_mem_bytes = _get_total_shared_memory()
+
+            (
+                ray_worker_heap_mem_bytes,
+                ray_worker_object_store_bytes,
+            ) = _calc_mem_per_ray_worker(
+                num_task_slots,
+                physical_mem_bytes,
+                shared_mem_bytes,
+                heap_to_object_store_ratio,
+            )
+            return ray_worker_heap_mem_bytes, ray_worker_object_store_bytes, None
+        except Exception as e:
+            return -1, -1, repr(e)
+
+    # Running memory inference routine on spark executor side since the spark worker nodes may
+    # have a different machine configuration compared to the spark driver node.
+    inferred_ray_worker_heap_mem_bytes, inferred_ray_worker_object_store_bytes, err = (
+        spark.sparkContext.parallelize([1], 1).map(mapper).collect()[0]
+    )
+
+    if err is not None:
+        raise RuntimeError(
+            f"Inferring ray worker available memory failed, error: {err}"
+        )
+    return inferred_ray_worker_heap_mem_bytes, inferred_ray_worker_object_store_bytes
+
+
+def get_spark_task_assigned_physical_gpus(gpu_addr_list):
+    if "CUDA_VISIBLE_DEVICES" in os.environ:
+        visible_cuda_dev_list = [
+            int(dev.strip()) for dev in os.environ["CUDA_VISIBLE_DEVICES"].split(",")
+        ]
+        return [visible_cuda_dev_list[addr] for addr in gpu_addr_list]
+    else:
+        return gpu_addr_list
+
+
+def _port_acquisition_delay():
+    """
+    Use a file lock to delay the worker processes to ensure that port acquisition does not
+    create a resource contention issue due to a race condition
+
+    Returns: None
+    """
+
+    def acquire_lock(file):
+        mode = os.O_RDWR | os.O_CREAT | os.O_TRUNC
+        file_ref = os.open(file, mode)
+        locked_file_ref = None
+        # Allow for retrying getting a file lock a maximum number of seconds
+        max_lock_iter = 60 * 2
+        for _ in range(max_lock_iter):
+            try:
+                fcntl.flock(file_ref, fcntl.LOCK_EX)
+            except (IOError, OSError):
+                pass
+            else:
+                locked_file_ref = file_ref
+                break
+            time.sleep(1.0)
+        if locked_file_ref is None:
+            os.close(file_ref)
+        return locked_file_ref
+
+    def release_lock(locked_file):
+        fcntl.flock(locked_file, fcntl.LOCK_UN)
+        os.close(locked_file)
+
+    file_lock = acquire_lock("/tmp/port_waiting_lock.lock")
```

Comment:
We can make file name more specific, like "/tmp/ray_on_spark_port_waiting_lock.lock"